pkg/exchange/ftx/exchange.go

```diff
@@ -179,7 +179,72 @@ func (e *Exchange) QueryAccountBalances(ctx context.Context) (types.BalanceMap,
 	return balances, nil
 }
 
+var supportedIntervals = map[types.Interval]int{
+	types.Interval1m:  1,
+	types.Interval5m:  5,
+	types.Interval15m: 15,
+	types.Interval1h:  60,
+	types.Interval1d:  60 * 24,
+	types.Interval3d:  60 * 24 * 3,
+}
+
+func (e *Exchange) SupportedInterval() map[types.Interval]int {
+	return supportedIntervals
+}
+
+func (e *Exchange) IsSupportedInterval(interval types.Interval) bool {
+	return isIntervalSupportedInKLine(interval)
+}
+
 func (e *Exchange) QueryKLines(ctx context.Context, symbol string, interval types.Interval, options types.KLineQueryOptions) ([]types.KLine, error) {
+	var klines []types.KLine
+	var since, until, current time.Time
+	if options.StartTime != nil {
+		since = *options.StartTime
+	}
+	if options.EndTime != nil {
+		until = *options.EndTime
+	} else {
+		until = time.Now()
+	}
+
+	current = until
+
+	for true {
+
+		endTime := current.Add(interval.Duration())
+		options.EndTime = &endTime
+		lines, err := e._queryKLines(ctx, symbol, interval, options)
+
+		if len(lines) == 0 {
```

Comment:
if len(lines) == 0 , but err is not nil ?

pre-check the error before this if block?

Reply:
ok